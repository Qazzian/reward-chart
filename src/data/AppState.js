import { combineReducers } from 'redux';
import * as fromCharts from './charts';
import * as fromEmotes from './emotes';

import {toEmoteDateString} from '../util/date';

// State

const AppState = combineReducers({
	charts: fromCharts.charts,
	emotes: fromEmotes.emotes,
});

export default AppState;


// Accessors

export const getAllCharts = (state) =>
	fromCharts.getAllCharts(state.charts);

export const getEmote = (state, emoteId) =>
	fromEmotes.getEmoteById(state.emotes, emoteId);

export const getEmoteList = (state, emoteIds) =>
	fromEmotes.getEmotesByIdList(state.emotes, emoteIds);


// Actions

export const addChart = fromCharts.addChart;

export function dispatchAddHappy(dispatch, chartId, date=(new Date())) {
	dispatchAddEmote(dispatch, chartId, date, fromEmotes.EMOTE_TYPES.HAPPY);
}

export function dispatchAddSad(dispatch, chartId, date=(new Date())) {
	dispatchAddEmote(dispatch, chartId, date, fromEmotes.EMOTE_TYPES.SAD);
}

function dispatchAddEmote(dispatch, chartId, date, emoteType) {

	const dateString = toEmoteDateString(date);

	const emoteId = fromEmotes.generateKey(chartId, dateString);
	dispatch(fromEmotes.addEmote(chartId, dateString, emoteType));
	dispatch(fromCharts.chartAddEmote(chartId, emoteId));
}
