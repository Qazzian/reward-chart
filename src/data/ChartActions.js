import {loadData, saveData} from './LocalStorage';

export const
	CHART_LOAD_REQUEST = 'CHART_LOAD_REQUEST',
	CHART_LOAD_RECEIVE = 'CHART_LOAD_RECEIVE',
	CHART_SAVE_REQUEST = 'CHART_SAVE_REQUEST',
	CHART_SAVE_COMPLETE = 'CHART_SAVE_COMPLETE',
	CHART_ADD = 'CHART_ADD',
	CHART_REMOVE = 'CHART_REMOVE',
	CHART_HAPPY = 'CHART_HAPPY',
	CHART_SAD = 'CHART_SAD';

export function requestChartLoad() {
	return {
		type: CHART_LOAD_REQUEST,
	}
}

export function receiveChartLoad(chartState) {
	return {
		type: CHART_LOAD_RECEIVE,
		data: chartState
	}
}

export function requestChartSave() {
	return {
		type: 'CHART_SAVE_REQUEST'
	}
}

export function completeChartSave() {
	return {
		type: 'CHART_SAVE_COMPLETE'
	}
}

export function addChart(chartData) {
	return {
		type: CHART_ADD,
		data: chartData,
	};
}

export function removeChart(chart) {
	return {
		type: CHART_REMOVE,
		data: chart,
	};
}

export function chartHappy(chart, date) {
	return {
		type: CHART_HAPPY,
		data: chart,
		date: date,
	};
}

export function chartSad(chart, date) {
	return {
		type: CHART_SAD,
		data: chart,
		date: date,
	};
}

export function fetchChartData() {
	return dispatch => {
		dispatch(requestChartLoad());
		return loadData()
			.then(response => dispatch(receiveChartLoad(response)));
	}
}

export function saveChartData(chartSate) {
	return dispatch => {
		dispatch(requestChartSave());
		return saveData(chartSate)
			.then(() => dispatch(completeChartSave()));
	}
}
