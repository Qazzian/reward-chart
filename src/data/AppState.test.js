import {createStore} from 'redux';

import AppState from './AppState';
import * as chartActions from "./ChartActions";


function createStoreFromState(initialState) {
	return createStore(AppState, initialState);
}


describe('AppState intigration tets', () => {
	
	it('deal with updates correctly', () => {
		const initState = {"charts":[{"name":"Chart 1"},{"name":"Chart 2"}]};
		const store = createStoreFromState(initState);
		store.dispatch(chartActions.chartHappy({name: "Chart 1"}, new Date('2018-12-10')));

		expect(store.getState()).toMatchSnapshot();


		
	});

});

