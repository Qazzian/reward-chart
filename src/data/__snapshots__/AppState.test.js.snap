// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppState intigration tets deal with updates correctly 1`] = `
Object {
  "charts": Array [
    Object {
      "emotes": Array [
        Object {
          "date": 2018-12-10T00:00:00.000Z,
          "emote": "HAPPY",
        },
      ],
      "name": "Chart 1",
    },
    Object {
      "name": "Chart 2",
    },
  ],
}
`;
